File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_VisualBasic/SolverGenerator/Techniques/NakedSubsetTechnique.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: NakedSubsetTecnnique.vb
'
'  Description: Implements the Sudoku naked subset technique.
' 
'--------------------------------------------------------------------------

Imports Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections

Namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
	''' <summary>Implements the naked pair elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class NakedPairTechnique
		Inherits NakedSubsetTechnique
		''' <summary>Initializes the technique.</summary>
		Public Sub New()
			MyBase.New(2)
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 5
			End Get
		End Property
	End Class

	''' <summary>Implements the naked triplet elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class NakedTripletTechnique
		Inherits NakedSubsetTechnique
		''' <summary>Initializes the technique.</summary>
		Public Sub New()
			MyBase.New(3)
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 7
			End Get
		End Property
	End Class

	''' <summary>Implements the naked quad elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class NakedQuadTechnique
		Inherits NakedSubsetTechnique
		''' <summary>Initializes the technique.</summary>
		Public Sub New()
			MyBase.New(4)
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 9
			End Get
		End Property
	End Class

	''' <summary>Implements the naked subset elimination technique.</summary>
	<Serializable>
	Public MustInherit Class NakedSubsetTechnique
		Inherits EliminationTechnique
		''' <summary>The size of the subset to evaluate.</summary>
		Private _subsetSize As Integer
		Private _foundLocations() As Integer

		''' <summary>Initialize the technique.</summary>
		''' <param name="subsetSize">The size of the subset to evaluate.</param>
		Protected Sub New(ByVal subsetSize As Integer)
			If subsetSize < 2 Then
				Throw New ArgumentOutOfRangeException("subsetSize")
			End If
			_subsetSize = subsetSize
			_foundLocations = New Integer(_subsetSize - 1){} ' optimization, rather than doing it in each call to EliminateNakedSubsets
		End Sub

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
            Dim nst = TryCast(obj, NakedSubsetTechnique)
			If nst Is Nothing Then
				Return False
			End If
			Return _subsetSize = nst._subsetSize
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Me.GetType().GetHashCode()
		End Function

		Public Overrides Function Clone() As EliminationTechnique
            Dim nst = CType(MyBase.Clone(), NakedSubsetTechnique)
			nst._foundLocations = New Integer(_subsetSize - 1){}
			Return nst
		End Function

		''' <summary>Performs naked subset elimination on one dimension of possible numbers.</summary>
		''' <param name="possibleNumbers">The row/column/box to analyze.</param>
		''' <returns>The number of changes that were made to the possible numbers.</returns>
        Private Function EliminateNakedSubsets(ByVal possibleNumbers() As FastBitArray, ByVal exitEarlyWhenSoleFound As Boolean,
            <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Integer
            Dim changesMade = 0
            exitedEarly = False
            Dim foundLocations() = _foundLocations ' optimization, rather than allocating each time

            For i = 0 To possibleNumbers.Length - 1
                If possibleNumbers(i).CountSet = _subsetSize Then
                    foundLocations(0) = i
                    Dim toMatchValues() = possibleNumbers(i).GetSetBits()
                    Dim matchesFound = 0
                    For j = i + 1 To possibleNumbers.Length - 1
                        If possibleNumbers(j).CountSet = _subsetSize Then
                            Dim foundMatch = CompareValues(toMatchValues, possibleNumbers(j).GetSetBits())
                            If foundMatch Then
                                matchesFound += 1
                                foundLocations(matchesFound) = j
                                If matchesFound = _subsetSize - 1 Then
                                    For k = 0 To possibleNumbers.Length - 1
                                        If Array.BinarySearch(foundLocations, k) < 0 Then
                                            For Each eliminatedPossible In toMatchValues
                                                If possibleNumbers(k)(eliminatedPossible) Then
                                                    changesMade += 1
                                                    possibleNumbers(k)(eliminatedPossible) = False
                                                End If
                                            Next eliminatedPossible
                                            If exitEarlyWhenSoleFound AndAlso possibleNumbers(k).CountSet = 1 Then
                                                exitedEarly = True
                                                Return changesMade
                                            End If
                                        End If
                                    Next k
                                    Exit For
                                End If
                            End If
                        End If
                    Next j
                End If
            Next i

            Return changesMade
        End Function

		''' <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		''' <param name="state">The puzzle state.</param>
		''' <param name="possibleNumbers">The previously computed possible numbers.</param>
		''' <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		''' <returns>Whether more changes may be possible based on changes made during this execution.</returns>
        Friend Overrides Function Execute(ByVal state As PuzzleState, ByVal exitEarlyWhenSoleFound As Boolean,
                        ByVal possibleNumbers()() As FastBitArray, <System.Runtime.InteropServices.Out()> ByRef numberOfChanges As Integer,
                        <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Boolean
            numberOfChanges = 0
            exitedEarly = False
            Dim arrays(state.GridSize - 1) As FastBitArray

            For i = 0 To state.GridSize - 1
                GetRowPossibleNumbers(possibleNumbers, i, arrays)
                numberOfChanges += EliminateNakedSubsets(arrays, exitEarlyWhenSoleFound, exitedEarly)
                If exitedEarly Then
                    Return False
                End If

                GetColumnPossibleNumbers(possibleNumbers, i, arrays)
                numberOfChanges += EliminateNakedSubsets(arrays, exitEarlyWhenSoleFound, exitedEarly)
                If exitedEarly Then
                    Return False
                End If

                GetBoxPossibleNumbers(state, possibleNumbers, i, arrays)
                numberOfChanges += EliminateNakedSubsets(arrays, exitEarlyWhenSoleFound, exitedEarly)
                If exitedEarly Then
                    Return False
                End If
            Next i

            Return numberOfChanges <> 0
        End Function
	End Class
End Namespace

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_CSharp/SolverGenerator/Techniques/NakedSubsetTechnique.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: NakedSubsetTecnnique.cs
//
//  Description: Implements the Sudoku naked subset technique.
// 
//--------------------------------------------------------------------------

using System;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
{
	/// <summary>Implements the naked pair elimination technique.</summary>
	[Serializable]
    public sealed class NakedPairTechnique : NakedSubsetTechnique
	{
		/// <summary>Initializes the technique.</summary>
		public NakedPairTechnique() : base(2) {}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 5; } }
	}

	/// <summary>Implements the naked triplet elimination technique.</summary>
	[Serializable]
    public sealed class NakedTripletTechnique : NakedSubsetTechnique
	{
		/// <summary>Initializes the technique.</summary>
		public NakedTripletTechnique() : base(3) {}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 7; } }
	}

	/// <summary>Implements the naked quad elimination technique.</summary>
	[Serializable]
    public sealed class NakedQuadTechnique : NakedSubsetTechnique
	{
		/// <summary>Initializes the technique.</summary>
		public NakedQuadTechnique() : base(4) {}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 9; } }
	}

	/// <summary>Implements the naked subset elimination technique.</summary>
	[Serializable]
    public abstract class NakedSubsetTechnique : EliminationTechnique
	{
		/// <summary>The size of the subset to evaluate.</summary>
		private int _subsetSize;
		private int [] _foundLocations;

		/// <summary>Initialize the technique.</summary>
		/// <param name="subsetSize">The size of the subset to evaluate.</param>
		protected NakedSubsetTechnique(int subsetSize)
		{
			if (subsetSize < 2) throw new ArgumentOutOfRangeException("subsetSize");
			_subsetSize = subsetSize;
			_foundLocations = new int[_subsetSize]; // optimization, rather than doing it in each call to EliminateNakedSubsets
		}

        public override bool Equals(object obj)
        {
            NakedSubsetTechnique nst = obj as NakedSubsetTechnique;
            if (nst == null) return false;
            return _subsetSize == nst._subsetSize;
        }

        public override int GetHashCode()
        {
            return GetType().GetHashCode();
        }

        public override EliminationTechnique Clone()
        {
            NakedSubsetTechnique nst = (NakedSubsetTechnique)base.Clone();
            nst._foundLocations = new int[_subsetSize];
            return nst;
        }

		/// <summary>Performs naked subset elimination on one dimension of possible numbers.</summary>
		/// <param name="possibleNumbers">The row/column/box to analyze.</param>
		/// <returns>The number of changes that were made to the possible numbers.</returns>
		private int EliminateNakedSubsets(FastBitArray [] possibleNumbers,
			bool exitEarlyWhenSoleFound, out bool exitedEarly)
		{
			int changesMade = 0;
			exitedEarly = false;
			int [] foundLocations = _foundLocations; // optimization, rather than allocating each time

			for(int i=0; i<possibleNumbers.Length; i++)
			{
				if (possibleNumbers[i].CountSet == _subsetSize)
				{
					foundLocations[0] = i;
					int [] toMatchValues = possibleNumbers[i].GetSetBits();
					int matchesFound = 0;
					for(int j=i+1; j<possibleNumbers.Length; j++)
					{
						if (possibleNumbers[j].CountSet == _subsetSize)
						{
							bool foundMatch = CompareValues(toMatchValues, possibleNumbers[j].GetSetBits());
							if (foundMatch)
							{
								foundLocations[++matchesFound] = j;
								if (matchesFound == _subsetSize-1)
								{
									for(int k=0; k<possibleNumbers.Length; k++)
									{
										if (Array.BinarySearch(foundLocations, k) < 0)
										{
											foreach(int eliminatedPossible in toMatchValues)
											{
												if (possibleNumbers[k][eliminatedPossible])
												{
													changesMade++;
													possibleNumbers[k][eliminatedPossible] = false;
												}
											}
											if (exitEarlyWhenSoleFound &&
												possibleNumbers[k].CountSet == 1)
											{
												exitedEarly = true;
												return changesMade;
											}
										}
									}
									break;
								}
							}
						}
					}
				}
			}

			return changesMade;
		}

		/// <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		/// <param name="state">The puzzle state.</param>
		/// <param name="possibleNumbers">The previously computed possible numbers.</param>
		/// <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		/// <returns>Whether more changes may be possible based on changes made during this execution.</returns>
		internal override bool Execute(
			PuzzleState state, bool exitEarlyWhenSoleFound, 
			FastBitArray[][] possibleNumbers, out int numberOfChanges, out bool exitedEarly)
		{
			numberOfChanges = 0;
			exitedEarly = false;
			FastBitArray [] arrays = new FastBitArray[state.GridSize];

			for(int i=0; i<state.GridSize; i++)
			{
				GetRowPossibleNumbers(possibleNumbers, i, arrays);
				numberOfChanges += EliminateNakedSubsets(arrays, exitEarlyWhenSoleFound, out exitedEarly);
				if (exitedEarly) return false;

				GetColumnPossibleNumbers(possibleNumbers, i, arrays);
				numberOfChanges += EliminateNakedSubsets(arrays, exitEarlyWhenSoleFound, out exitedEarly);
				if (exitedEarly) return false;

				GetBoxPossibleNumbers(state, possibleNumbers, i, arrays);
				numberOfChanges += EliminateNakedSubsets(arrays, exitEarlyWhenSoleFound, out exitedEarly);
				if (exitedEarly) return false;
			}

			return numberOfChanges != 0;
		}
	}
}